### PR TITLE
pull Checkpoints out of #2289

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
current Bank checkpoint code supports only a single stack of checkpoints, cannot support 2 simultaneous forks

#### Summary of Changes
implement Checkpoints, a directed acyclic graph, uses u64 (probably block index) as key for a checkpoint

### CODE REVIEW

**Comment by garious** [comment]
Looks like a sloppy specialization of a DAG. Have you looked at:

https://docs.rs/petgraph/0.4.13/petgraph/

**Comment by rob-solana** [comment]
> Looks like a sloppy specialization of a DAG. Have you looked at:
> 
> https://docs.rs/petgraph/0.4.13/petgraph/
I've just had a look, and I don't know if using it would be less code.  Maybe I'd be able to copy and paste some generic DAG code for trunk() and collect(), but would that be better?

File: src/checkpoints.rs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                fork = *parent;
+            } else {
+                break;
+            }
+        }
+        rv
+    }
+    /// given a maximum depth, find the highest parent shared by all
+    ///  forks
+    pub fn trunk(&self, num: usize) -> Option<u64> {
```
Comment (aeyakovenko) [inline src/checkpoints.rs]: why not a method that does
```
/// returns the trunks, and an inverse dag
fn invert(&self) -> (HashSet<u64>, HashMap<u64, HashSet<u64>>)
```

**Comment by garious** [comment]
It's not clear what you're trying to do here (not integrated into anything, no issue referenced, and there's no PR description). You're adding what appears to be a fairly generic library awkwardly using the same name as the existing Checkpoint trait and specialized for `u64` IDs.

File: src/checkpoints.rs (hunk 1 above)
Comment (rob-solana) [reply src/checkpoints.rs]: and then?  probably what you've written is more along the lines of what you want for bank_state?

**Comment by rob-solana** [comment]
> It's not clear what you're trying to do here (not integrated into anything, no issue referenced, and there's no PR description). You're adding what appears to be a fairly generic library awkwardly using the same name as the existing Checkpoint trait and specialized for `u64` IDs.

new trait name
this piece of #2289 is easily separable and is teaching me something about that code

File: src/checkpoints.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+        // 3->2->1
+        // 4->2->1
+        checkpoints.store(1, 0, 0);
+        checkpoints.store(2, 0, 1);
+        checkpoints.store(3, 0, 2);
+        checkpoints.store(4, 0, 2);
+
+        let _trunks: Vec<_> = checkpoints.invert().0.into_iter().collect();
+        // TODO: returns [0, 2] ??
```
Comment (rob-solana) [inline src/checkpoints.rs]: @aeyakovenko is this the expected behavior?

File: src/checkpoints.rs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+        // 30->20->10
+        // 40->20->10
+        checkpoints.store(10, 0, 0);
+        checkpoints.store(20, 0, 10);
+        checkpoints.store(30, 0, 20);
+        checkpoints.store(40, 0, 20);
+
+        let _trunks: Vec<_> = checkpoints.invert().0.into_iter().collect();
+        // TODO: returns [0, 2, 10, 20, 1] ??
```
Comment (rob-solana) [inline src/checkpoints.rs]: @aeyakovenko is this the expected behavior?  I don't understand how 1 is now part of trunks

**Comment by rob-solana** [comment]
updated description, tests

**Pull request closed without merging**